First step toward statistics in ORC files with chunked writes

hyperbolic2346 opened the pull request:
We are making the changes necessary to fix issue #5826. This is the first of those changes and is a refactor of the way statistics are captured to better support chunked writing. The next change will include things like multiple data pointers passed to the kernels and storage between calls to write.

hyperbolic2346 commented:
I hadn't received any review comments, so I thought I would merge in my latest work here instead of making a new PR. This should be all the structure changes necessary for chunked writes. I will make another PR to enable them and persist the data necessary there.

vuule reviewed (changes requested):
partial review - potential bug fix, and a few nit picks

hyperbolic2346 commented on cpp/src/io/orc/writer_impl.cu:
```diff
@@ -1186,20 +1196,93 @@ writer::impl::encoded_statistics writer::impl::gather_statistic_blobs(
   }
-  return {std::move(rowgroup_blobs), std::move(stripe_blobs), std::move(file_blobs)};
+
+  return {std::move(stripe_blobs), std::move(file_blobs)};
```
It was, but since row groups are needed earlier than when we call `finish_statistics_blobs` it isn't in this structure anymore. Row group information is consumed in write still.

vuule replied:
Missed that part :( Might be good to rename `encoded_statistics`, since it only includes the ones that are written in the file footer.

hyperbolic2346 replied:
A great idea, I'm thinking `encoded_footer_statistics`

hyperbolic2346 commented on cpp/src/io/orc/writer_impl.hpp:
```diff
@@ -284,24 +284,61 @@ class writer::impl {
+   * @param incoming_stats intermediate statistics returned from `gather_statistic_blobs`
+   * @return The encoded statistic blobs
+   */
+  encoded_statistics finish_statistic_blobs(int const num_columns,
+                                            int const num_stripes,
```
We can. This was created before I conceded storing off the stats_desc inside the intermediate stats. I'll update that.

hyperbolic2346 replied:
Actually we can't remove `num_stripes` because we do have the number of stripe blobs, we don't know how many stripes unless there is some math there to compute it. I've removed `num_columns` and left `num_stripes` for now.

hyperbolic2346 commented on cpp/src/io/orc/writer_impl.hpp:
```diff
@@ -284,24 +284,61 @@ class writer::impl {
     hostdevice_2dvector<gpu::encoder_chunk_streams>* enc_streams,
     hostdevice_2dvector<gpu::StripeStream>* strm_desc);
 
+  struct intermediate_statistics {
```
Done

vuule commented:
rerun tests

hyperbolic2346 commented:
Fixed the dangling pointer issue that was tripping up python tests and refactored the code some. I think that the finish work can move back into close, but I'll leave that for the next PR so as to not complicate this one anymore.

vuule commented on cpp/src/io/orc/writer_impl.hpp:
```diff
@@ -286,24 +286,82 @@ class writer::impl {
+    std::vector<hostdevice_vector<statistics_merge_group>> stripe_stat_merge;
+    std::vector<statistics_dtype> stats_dtypes;
+    std::vector<data_type> col_types;
+    int num_rows;
```
just to be on the safe side (otherwise we need a ctor).
```suggestion
    int num_rows = 0;
```

vuule commented on cpp/src/io/orc/writer_impl.hpp:
```diff
@@ -286,24 +286,82 @@ class writer::impl {
+      stripe_stat_chunks.clear();
+      stripe_stat_merge.clear();
+      stats_dtypes.clear();
+      col_types.clear();
```
if "empty `persisted_statistics`" implies zero rows. If it does not, we need a constructor that sets `num_rows`.
```suggestion
      col_types.clear();
      num_rows = 0;
```

vuule reviewed (approved):
looks good! Few very minor suggestions.

hyperbolic2346 commented on cpp/src/io/statistics/statistics.cuh:
```diff
@@ -107,7 +107,8 @@ struct statistics_group {
 
 struct statistics_merge_group {
-  const stats_column_desc* col;  //!< Column information
+  data_type col_type;            //!< Column information
```
I was torn here because the original variable inside `stats_column_desc` is named col_type. I agree though, we should change this to be consistent.

bdice commented on cpp/src/io/orc/writer_impl.cu:
```diff
@@ -55,7 +55,6 @@
 #include <utility>
 
 #include <cuda/std/limits>
-
```
Can we re-add this blank line? It keeps the include sections separated.

hyperbolic2346 replied:
unintentional, I'll fix it.

bdice commented on cpp/src/io/orc/writer_impl.cu:
```diff
@@ -1073,23 +1073,49 @@ void set_stat_desc_leaf_cols(device_span<orc_column_device_view const> columns,
+  // figure out the buffer size needed for protobuf format
+  gpu::orc_init_statistics_buffersize(
+    stats_merge_groups.device_ptr(), stat_chunks.data(), num_stat_blobs, stream);
+  stats_merge_groups.device_to_host(stream, true);
```
It looks like only `stats_merge_gropus[num_stat_blobs - 1]` needs to be read from this device-to-host copy. Is it possible to copy only one value (or even just the `start_chunk` and `num_chunks`) instead of the entire vector?

hyperbolic2346 replied:
There is a good deal of device to host and back. I kept having to reel myself in here as I didn't want to drastically alter this. There isn't a way to get a `rmm::device_buffer` element to the host and `rmm::device_uvector` requires removing the default constructor. So for now I'm going to lift the element function from `rmm::device_uvector`, but I don't like that either. I guess it is a step in the right direction...

bdice commented on cpp/src/io/orc/writer_impl.cu:
```diff
@@ -1121,96 +1147,163 @@ writer::impl::encoded_statistics writer::impl::gather_statistic_blobs(
       grp.start_chunk =
         static_cast<uint32_t>(column.index() * segmentation.num_rowgroups() + stripe.first);
       grp.num_chunks = stripe.size;
       for (auto rg_idx_it = stripe.cbegin(); rg_idx_it < stripe.cend(); ++rg_idx_it) {
```
The pattern `!= end_iterator` is slightly better because it doesn't require a random access iterator and doesn't require a check for "before" ordering.
```suggestion
      for (auto rg_idx_it = stripe.cbegin(); rg_idx_it != stripe.cend(); ++rg_idx_it) {
```

bdice commented on cpp/src/io/orc/writer_impl.cu:
```diff
@@ -1121,96 +1147,163 @@ writer::impl::encoded_statistics writer::impl::gather_statistic_blobs(
-  auto rowgroup_stat_chunks = stat_chunks.data();
-  auto stripe_stat_chunks   = rowgroup_stat_chunks + num_rowgroup_blobs;
-  auto file_stat_chunks     = stripe_stat_chunks + num_stripe_blobs;
+  // rowgroup stat chunks are written out in each stripe stripe and file-level chunks are written in
+  // the footer. to prevent persisting the rowgroup stat chunks across multiple calls to write in a
+  // chunked write situation, these allocations are split up so stripe data can persist until the
+  // footer write
```
Tried to improve wording here.
```suggestion
  // The rowgroup stat chunks are written out in each stripe and file-level chunks are written in
  // the footer. To prevent persisting the rowgroup stat chunks across multiple write calls in a
  // chunked write situation, these allocations are split up so stripe data can persist until the
  // footer is written.
```

hyperbolic2346 replied:
I'm going to commit this and then edit it a bit, would be nice to re-review. The stripe stripe was actually different thoughts. Rowgroup stat chunks are written out in each stripe. And then stripe and file-level chunks are written in the footer. I'll start with yours and attempt to make this more clear.

bdice commented on cpp/src/io/orc/writer_impl.cu:
```diff
@@ -1121,96 +1147,163 @@ writer::impl::encoded_statistics writer::impl::gather_statistic_blobs(
-  stat_merge.device_to_host(stream, true);
+  // with chunked writes, the orc table can be deallocated between write calls.
+  // This forces our hand to encode row groups and stripes only in this stage and further
+  // we have to persist any data from the table that we need, min/max strings.
```
I'm not sure if I understand this line. What does "min/max strings" mean at the end? Is that an example of the type of data that must be kept until the end of the chunked write? Are statistics always stored as strings?

hyperbolic2346 replied:
We are saving the strings stored in the statistics chunks, which are the minimum length string and the maximum length string. I'll clean this up as well, thanks for pointing it out.

hyperbolic2346 replied:
Reworded, please advise.

bdice commented on cpp/src/io/orc/writer_impl.cu:
```diff
@@ -1121,96 +1147,163 @@ writer::impl::encoded_statistics writer::impl::gather_statistic_blobs(
+    }
+
+    auto d_file_stats_merge = stats_merge.device_ptr(num_stripe_blobs);
+    cudaMemcpyAsync(d_file_stats_merge,
```
It looks like this is using a raw `cudaMemcpyAsync` on a `hostdevice_vector` object. Can we use its built-in methods to avoid a raw `cudaMemcpyAsync` call?

hyperbolic2346 replied:
Nice catch!

bdice reviewed (commented):
Some comments attached. I'd like another person to be the second approver for this PR since I'm not very familiar with this code and it requires a fairly intimate knowledge of the format to verify correctness.

vuule commented:
> Some comments attached. I'd like another person to be the second approver for this PR since I'm not very familiar with this code and it requires a fairly intimate knowledge of the format to verify correctness.

Thank you for the review!
Well, Devavret is out, so I don't think we're getting a second reviewer who is very familiar with statistics :D FWIW, Mike went over the semantics with Devavret as well, so IMO we should be good with a non-cuIO second review.

mythrocks reviewed (approved):
Sorry for the delay. It takes me a while to grok code.

I'm generally :+1: apart from some minor questions and nitpicks.

hyperbolic2346 commented:
@gpucibot merge

mythrocks commented on cpp/src/io/orc/writer_impl.hpp:
```diff
@@ -284,24 +284,61 @@ class writer::impl {
+   * @brief Returns column statistics encoded in ORC protobuf format.
+   *
+   * @param stats_freq Frequence of statistics to be included in the output file
+   * @param orc_table Table information to be written
```
Argh. Sorry, this was in the old code. (These were in my review of your initial draft. I honestly thought I'd deleted this comment.)
